Add JCC rest api calls

viktorvanwijk opened the pull request:
Partly closes #5696

[skip: e2e]

**Changes**

Implemented the following API calls:
- get_available_products
- get_locations
- get_dates
- get_times

**Checklist**

Check off the items that are completed or not relevant.

- Impact on features

  - [x] Checked copying a form
  - [x] Checked import/export of a form
  - [x] Config checks in the configuration overview admin page
  - [x] Problem detection in the admin email digest is handled

- Dockerfile/scripts

  - [x] Updated the Dockerfile with the necessary scripts from the `./bin` folder

- Commit hygiene

  - [x] Commit messages refer to the relevant Github issue
  - [x] Commit messages explain the "why" of change, not the how

viktorvanwijk committed:
:sparkles: [#5696] Implement JCC API calls

- Locations
- Products (they are called 'activities' in JCC)
- Available dates
- Available times

viktorvanwijk committed:
:sparkles: [#5696] Add config check to JCC Rest plugin

vaszig commented on src/openforms/appointments/api/serializers.py:
```diff
@@ -55,6 +55,9 @@ class ProductSerializer(serializers.Serializer):
         label=_("identifier"), help_text=_("ID of the product")
     )
     name = serializers.CharField(label=_("name"), help_text=_("Product name"))
+    description = serializers.CharField(
```
I guess this can be removed as it is part of my PR specifically for the description (backend and frontend).

vaszig commented on src/openforms/appointments/contrib/jcc_rest/tests/test_plugin.py:
```diff
@@ -0,0 +1,318 @@
+
+        self.assertEqual(len(products), 26)
+        # Select product with a description
+        product = products[-2]
```
Are we sure about the order of the products/activities that the API returns?I do not see something like a fixed order. Maybe we could grab the product from the id?Of course many other things can change in the test instance in the future but just thinking.

vaszig commented on src/openforms/appointments/contrib/jcc_rest/tests/test_plugin.py:
```diff
@@ -0,0 +1,318 @@
+
+        mock_client.return_value = Client()
+
+        plugin = JccRestPlugin("jcc")
```
```suggestion
        plugin = JccRestPlugin("jcc_rest")
```

vaszig commented on src/openforms/appointments/contrib/jcc_rest/client.py:
```diff
@@ -0,0 +1,248 @@
+
+class Activity(TypedDict):
+    id: str
+    number: NotRequired[int]  # will be replaced by id eventually
```
I don't see that as optional in the API spec, do I miss something?

vaszig commented on src/openforms/appointments/contrib/jcc_rest/client.py:
```diff
@@ -0,0 +1,248 @@
+    name: NotRequired[str]
+
+
+class Address(TypedDict):
```
What about the `revision`? Why is it not present in any related typedDict?

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/client.py:
```diff
@@ -0,0 +1,248 @@
+
+    def get_activity_list_for_appointment(
+        self, location_id: str
+    ) -> Collection[Activity]:
```
Yeah good point. I guess "list" in the documentation and endpoint name(s) implies it should be a sequence, which means we can actually trust the order in the tests

viktorvanwijk commented on src/openforms/appointments/contrib/jcc_rest/tests/test_plugin.py:
```diff
@@ -0,0 +1,318 @@
+        }
+
+        with self.subTest("No information"):
+            location = self.plugin._create_location(raw_location)  # pyright: ignore[reportAttributeAccessIssue]
```
Had another look, and I guess there is a way to do it more efficiently. Originally thought I had to create four separate tests for it, but it's possible to just modify the return value multiple times. Will change it :+1:

viktorvanwijk committed:
:ok_hand: [#5696] Process PR feedback

* Move typed dicts to separate file.
* Mark `Activity.number` and `Location.locationNumber` as required, according to the API docs.
* Use sequence instead of collection for return values of API calls. The documentation
  talks specifically about lists and arrays.
* Change the query parameter list to a dict - looks a bit cleaner.
* Use the complete flow when testing the address processing, by mocking the response. The example
  data from JCC does not contain all combinations for us to test.
* Combine activity IDs and amounts into one function argument. This makes the assert on the
  sequence lengths explicit.

vaszig commented on src/openforms/appointments/contrib/jcc_rest/client.py:
```diff
@@ -150,22 +72,24 @@ def get_activity_list_for_appointment(
 
     def get_additional_activity_list_for_appointment(
         self, activity_ids: Collection[str], location_id: str
```
I see some are declared as `Collection` and some as `Sequence` below. Is there a specific difference and we do it that way?

vaszig reviewed (commented):
Looks good to me!A couple of small remarks.
